### DISCUSSION TITLE
Data structure

### DISCUSSION THREAD

**Discussion by roaldarbol** [VioletteChiara/AnimalTA#20] [General]
Hi @VioletteChiara! 

I'm writing an R package for analysis of animal movement from an array of different sources (pose estimation, centroid tracking, trackball/treadmill) and would love to support AnimalTA. I write reader functions for each data source, along with some validator functions that ensures that the files *really* are in the format that the user thinks they are. I received some single-animal data from a friend that uses AnimalTA (I haven't had time to try AnimalTA myself), and I'm wondering whether that is also valid for multi-animal data? Or are the data headers different when there are multiple individuals? So whether it is something like `Frame`, `Time`, `ID`, `x`, `y` or rather `Frame`, `Time`, `x_1`, `y_1`, `x_2`, `y_2` when there are two individuals?

Additionally, can I expect other column values inside an AnimalTA data frame? 

The package is currently called `trackballr`, but will change name to `animovement` one of the following days. I collaborate with the team making the `movement` Python package, so once I've gotten the reader and validators implemented it would also be a smaller task for them to support it. :-)

All the best, Mikkel 😊

**Comment by roaldarbol** [comment]
Oh, and I can see that the x/y columns in the dataset I received are called `X_Arena0_Ind0`/`Y_Arena0_Ind0`. Should I take that to mean that there can be multiple arenas and individuals, and that they would be added as extra columns? Maybe if you have some sample data of a small range of use-cases you can share that with me? :-)

**Reply by VioletteChiara** [reply]
Hi Mikkel!

Thanks a lot for thinking of including AnimalTA reader in your package! I will try my best to assist you and explain everything you want :)

There are three kind of coordinates file that AnimalTA is creating:

### 1. The raw coordinates: 
These data are the ones you will get immediately after the tracking: they do not include the corrections made manually by the user, all coordinates are expressed in pixels.
      o **If the user is tracking a known number of targets**: the structure is the one your friend sent you. As you guessed, you have two columns per individual + Frame and Time. The columns are: Frame, Time, X_ArenaA_IndI, Y_ArenaA_IndI (as you imagine, X is X coordinates, Y Y coordinates). AnimalTA can be used for multi-individual and multi-arenas trackings. The ArenaN indicates in which arena the target is (N coding for the arena ID, ranging from 0 to the number of arenas). The IndI indicates the ID of the target inside the considered arena (I coding for the individual ID, ranging from 0 to the number of individuals inside the considered arena). For instance, if you have one individual in one arena you get: Frame, Time, X_Arena0_Ind0, Y_Arena0_Ind0. If you have two arenas with one individual in each you get: Frame, Time, X_Arena0_Ind0, Y_Arena0_Ind0, X_Arena1_Ind0, Y_Arena1_Ind0. If you have one arena with 1 individual and another one with two individuals, you will have: Frame, Time, X_Arena0_Ind0, Y_Arena0_Ind0, X_Arena1_Ind0, Y_Arena1_Ind0, X_Arena1_Ind1, Y_Arena1_Ind1. All frames are represented and missing data are reported as "NA".

  o **If the user is tracking a variable number of targets**: This case means that the targets can enter and leave the tracking area (see https://x.com/i/status/1797906483902136510 for illustration). In that case, the number of targets increases very fast, so the system of storing data is a bit different. You will have a fixed number of columns: Frame, Time, Arena, Ind, X, Y. Before a target is seen for the first time or after it lefts the video, no data relative to this target is reported (no NA). However if a target is seen then lost while we still expect it to be inside the tracking arena, the coordinates will be reported as NA. 

### 2. The corrected coordinates:
These data are the ones you will get after the corrections made manually by the user: the structure is exactly similar to the one explained above

### 3. The detailed coordinates:
This last data frame is the one including the coordinates of the target and a lot of the information the user added after the tracking/during corrections. It is very different from the others as there is one file per individual. In this file the coordinates are not expressed in pixels anymore but in the international unit chosen by the user (if a scale is defined, if not we keep expressing in pixels). 
The basic columns are: Frame, X, Y, Speed, Moving, Distance. 
If the user choose to smooth the data for the analysis, these coordinates are the result after smoothing and the column names become: Frame, X_smoothed, Y_smoothed, Speed, Moving, Distance.
If the user used the correction of perspective tool the coordinates are also transformed. 
Speed is simply the movement speed, moving is whether the target is moving or not according to a movement threshold defined by the user. Distance is the distance traveled between the two frames.
But the number of column can later increase according to the situation: if there is more than one individual in the arena, the program will add one column per other individual and indicates in this column the inter-individual distance (column name: Dist_to_Ind43, but the individual ID may change if the user gave it a name). The user can also add elements of interest such as points, segments, borders, or arenas. There will be one column per element of interest (associated with the target arena) indicating the distance between the target and the element (name of the column: Dist_to_Point_0, but the "Point_0" part will change if the user gives it a name or according to the kind of element it is).

You will find attached examples of different data frames. Please do not hesitate to tell me if you have more questions or if there is anything I can do to help. You can also contact me directly to my mail address (you can find it at http://vchiara.eu/). I can then send you many more files.

I hope all this is helpful!

Best regards,

Violette





[Detailed_data_1ind_per_arena.csv](https://github.
[Raw_coordinates_variable_number_targets_1arena.csv](https://github.com/user-attachments/files/16933520/Raw_coordinates_variable_number_targets_1arena.csv)
com/user-attachments/files/16933491/Detailed_data_1ind_per_arena.csv)
[Raw_coordinates_1_ind_per_arena_8_arenas.csv.csv](https://github.com/user-attachments/files/16933528/Raw_coordinates_1_ind_per_arena_8_arenas.csv.csv)

**Reply by roaldarbol** [reply]
Hi Violette!

Thanks for the thorough walkthrough, it is *really* helpful. The package (called *animovement*) is beginning to take shape, so you can find it at https://www.roald-arboel.com/animovement/. I will have to take some time to dig in properly in the coming month (I've made the appropriate issue at https://github.com/roaldarbol/animovement/issues/26), so for now I just have a few quick questions from my quick read-through:
- What is the unit of `speed`? I'm curious since you mention frames, but no other time-unit. 
- I had a look at the Twitter video - loving the feature! I assume (but just want to be 100% sure) that there's no recognition in case the same individual re-enters the tracking after having left it? You just increase id number by 1 every time a new track is initiated?

I might come back with more questions once I begin implementing, hope that's alright!
Best, Mikkel

**Reply by VioletteChiara** [reply]
Great, seems nice, I will have a look :)

- Speed is normally expressed in units per second. If the user has not added a scale, it is expressed in px/sec. If the user has specified a scale, it can be expressed in m/sec, cm/sec or mm/sec, depending on the user's choice.
- Yes indeed, there is no recognition. There is only one particular case: if the target leaves the arena but stays in the "entrance area", meaning it is still tracked, but the tracking data is not kept after the tracking process. In that particular situation, a target that re-enter will still be recognized as being the same as before. 

Do not hesitate if you have more questions!

**Reply by roaldarbol** [reply]
Does the user themselves specify the FPS then? It was more since there was no other temporal units than frames - how would it know how many frames there is in a second? (I'm quite sure both DLC and SLEAP only express frames, not seconds unfortunately). In which case, the only way to get the actual time is by reverse-engineering from the speed. Maybe it would be worth adding time as a column if FPS has been specified? Would make synchronisation with other data sources easier too. :-)

EDIT: I was reading your **Detailed coordinates** comment above, where it doesn't say Time - but I noticed it does in the previous comments, I think that's what confused me. Is Time always present in all exported data frames?

Ah, that makes sense! Good to know about that special case too! Is the reason that they're actually still being tracked in the entire frame, but you only keep observations inside the ROI? (That's just me being curious) :-)

**Reply by VioletteChiara** [reply]
The FPS is extracted from the file metadata (can also be modified in case of corrupted metadata), but the user can change it later (the program will then do a subsample of the frames).
![image](https://github.com/user-attachments/assets/9cb28edf-3d63-4df6-aa19-11a0b6e62fc3)

Yes, it's my bad, I wrote wrong in my previous comment (sorry). The time is expressed in Frames for raw and corrected coordinates, but in Seconds for the Detailed data (where the speed is also mentioned). 
To be honest, I thought that people would use the "Detailed data" for exportation rather than the raw or corrected data, but that does not seem to be the case. So it is a good idea. I can add one more "Time" column in the future. 

Regarding the entrance data, there are two kinds of ROIs: the arena itself, which is where the track data are kept, and the entrance area which is from where we expect the targets to enter the arena. So they are indeed actually tracked outside of the arena (main ROI) but not everywhere else, only in the secondary ROIS (entrance area).
![image](https://github.com/user-attachments/assets/6e199c9b-8c1a-4ed2-aba3-435c559f33c2)

**Reply by roaldarbol** [reply]
Ah, that makes a lot of sense, thanks again for the thorough replies, amazingly helpful! BTW, I really like your development approach which seems focused on features the users *ask for* as opposed to what you *think* users want - it's a breath of fresh air - makes me happy to recommend AnimalTA to others!

**Reply by VioletteChiara** [reply]
Hi! Yes it is indeed the way the program is evolving, I try to encourage users to give me feedback and suggestions and if those are feasible I implement them in the program.

**Reply by VioletteChiara** [reply]
And thanks for the recommendations!

**Reply by roaldarbol** [reply]
Hi again @VioletteChiara, hope you're well!
Now I'm finally getting around to implementing the reader! :-D

> Raw_coordinates_1_ind_per_arena_8_arenas.csv.csv

In this file there seems to be a lot of `.` in the numbers, e.g. `5.574.633.782` and the same goes for time. How should I interpret that? Is it only the first . that actually represents a comma?

Thanks!

**Reply by VioletteChiara** [reply]
Hi @roaldarbol!

Sorry for the delay in my answer, I had very busy days. It's great that you are progressing that fast!
I don't see the "." you mentioned except one for decimals. Here is a screenshot from the file when I open it:
![image](https://github.com/user-attachments/assets/c54e80de-a3b5-4590-93c6-f23b214f44ab)

I guess it comes from your excel settings. Could you please try to open the file with something like the notepad and see if you still have those dots you mentionned?

Best wishes!

Violette

**Reply by roaldarbol** [reply]
No worries - there's no urgency on my end. :-)

Unfortunately they are still there, even in a plain text editor:
![Screenshot 2024-11-04 at 15 43 49](https://github.com/user-attachments/assets/fba25d4a-f918-462a-a515-983341d1e066)

Can you try to download the file from here to see if you get the same? 

EDIT: I just re-downloaded the file and now they're no longer there! Possibly Excel has messed up the file for me - no worries, all good, sorry about that!

**Reply by VioletteChiara** [reply]
Great that you found the solution! I was completly lost as I had downloaded it from here just before ='D.

**Reply by roaldarbol** [reply]
I got a file from a colleague with the following headers: `Frame;Time;X_Arena0_Ind0;Y_Arena0_Ind0`. So this is a single arena I guess - so am I right in thinking that whenever a ROI has been drawn it ends up in this format? I'm just figuring out whether I should name the parameter `multi_arena` or `roi` to distinguish between the two file types. :-)

**Reply by VioletteChiara** [reply]
Hi! This seems to be a "raw coordinates" file with a single animal in a single arena. The notion of ROI is a bit confusing in AnimalTA because there are two kinds of ROIs. The first one is the "arena" kind, which defines where the targets can be (for instance if I record an ant inside a Petri dish, it will be the whole Petri dish). A file with  "Frame;Time;X_Arena0_Ind0;Y_Arena0_Ind0" columns can be either a one-arena or a no-arena video.  Indeed, by default, if no arena is indicated by the user the program will consider there is one big arena which includes the whole frame. 


If there is one arena with one individual in each: Frame;Time;X_Arena0_Ind0;Y_Arena0_Ind0
If there is one arena with two individuals inside: Frame;Time;X_Arena0_Ind0;Y_Arena0_Ind0;X_Arena0_Ind1;Y_Arena0_Ind1

If there are two arenas with one individual in each: Frame;Time;X_Arena0_Ind0;Y_Arena0_Ind0;X_Arena1_Ind0;Y_Arena1_Ind0
If there are two arenas with two individuals in each: Frame;Time;X_Arena0_Ind0;Y_Arena0_Ind0;X_Arena0_Ind1;Y_Arena0_Ind1;X_Arena1_Ind0;Y_Arena1_Ind0;X_Arena1_Ind1;Y_Arena1_Ind1


After, there is what is called inside the program the "elements of interest" which are areas the user is especially interested in for the analyses. If I take again my ant example: the 1cm border inside the Petri dish could be an element of interest, or if there is a patch of food, the region around it, etc. The thing is that the elements of interest are drawn after tracking and so are not present in the raw coordinates files.

The relationship between the elements of interest and the coordinates are available in the "Detailed data" files that I described in a previous answer: "
_3. The detailed coordinates:
This last data frame is the one including the coordinates of the target and a lot of the information the user added after the tracking/during corrections. It is very different from the others as there is one file per individual. In this file, the coordinates are not expressed in pixels anymore but in the international unit chosen by the user (if a scale is defined, if not we keep expressing them in pixels).
The basic columns are: Frame, X, Y, Speed, Moving, Distance.
If the user choose to smooth the data for the analysis, these coordinates are the result after smoothing and the column names become: Frame, X_smoothed, Y_smoothed, Speed, Moving, Distance.
If the user used the correction of perspective tool the coordinates are also transformed.
Speed is simply the movement speed, moving is whether the target is moving or not according to a movement threshold defined by the user. Distance is the distance traveled between the two frames.
But the number of column can later increase according to the situation: if there is more than one individual in the arena, the program will add one column per other individual and indicates in this column the inter-individual distance (column name: Dist_to_Ind43, but the individual ID may change if the user gave it a name). The user can also add elements of interest such as points, segments, borders, or arenas. There will be one column per element of interest (associated with the target arena) indicating the distance between the target and the element (name of the column: Dist_to_Point_0, but the "Point_0" part will change if the user gives it a name or according to the kind of element it is).
"_

**Reply by roaldarbol** [reply]
Ah okay, I see thanks! I think I will focus on supporting just the first type for now - I am planning to add similar features (e.g. "elements of interest"), so it can be used on all data from all sources. However, I already implemented it for the "Detailed" file with just the basic columns - but I imagine issues would arise if it has more complicated columns (such as `dist_to_point_0`) where I don't know where the point is. 

Anyway, I've finished implementing the reader and it's now available in v0.4.0 🥳 (and just for the record, the package is still in beta)😅

**Reply by roaldarbol** [reply]
The reference for the function is here: https://roald-arboel.com/animovement/reference/read_animalta.html
I'll change the parameter `with_roi` to `detailed` in a patch later today. 😊

**Reply by VioletteChiara** [reply]
Great! I will give it a try in the upcoming days :)
Thanks a lot for all your work!

**Reply by roaldarbol** [reply]
Sounds good! I made a patch now. Please let me know if something doesn't work well! There's a chance installation might be a bit tricky, and it would be great to know whether it is for you. :)
